Fix memory leak of resultingModuleIndex and handle g_object refs

kontura opened the pull request:
If we want to keep some modulemd objects (such as streams, profiles..)
even after unrefing `resultingModuleIndex` in `ModuleMetadata` destructor,
we have to increase references to them in their respective libdnf
counterparts, otherwise unreferencing `resultingModuleIndex` will 
unreference the whole tree of module metadata and it will all be freed.

This is to resolve: https://github.com/rpm-software-management/libdnf/issues/977

kontura committed:
Fix memory leak of resultingModuleIndex and handle g_object refs

If we want to keep some modulemd objects (such as streams, profiles..)
even after unrefing resultingModuleIndex in ModuleMetadata destructor,
we have to increase references to them in their respective libdnf
counterparts.

sbstnk commented:
I tried building a libdnf package for F32 with the patch applied, but the build fails because of a segfault in one of the tests:

```
test 1
    Start 1: test_cpp

1: Test command: /builddir/build/BUILD/libdnf-0.48.0/build-py3/tests/run_tests "DEPENDS" "run_tests" "COMMENT" "Running CPPUNIT tests..."
1: Environment variables: 
1:  LD_LIBRARY_PATH=/builddir/build/BUILD/libdnf-0.48.0/build-py3/libdnf
1: Test timeout computed to be: 10000000
1: .
1: (process:2250): GLib-GObject-CRITICAL **: 11:03:17.176: g_object_unref: assertion 'G_IS_OBJECT (object)' failed
1: 
1: (process:2250): libmodulemd-CRITICAL **: 11:03:17.180: modulemd_profile_get_name: assertion 'MODULEMD_IS_PROFILE (self)' failed
1: F
1: (process:2250): GLib-GObject-CRITICAL **: 11:03:17.180: g_object_unref: assertion 'G_IS_OBJECT (object)' failed
1: 
1: (process:2250): GLib-GObject-CRITICAL **: 11:03:17.180: g_object_unref: assertion 'G_IS_OBJECT (object)' failed
1: .
1: (process:2250): GLib-GObject-CRITICAL **: 11:03:17.184: g_object_unref: assertion 'G_IS_OBJECT (object)' failed
1: 
1: (process:2250): libmodulemd-CRITICAL **: 11:03:17.187: modulemd_profile_get_description: assertion 'MODULEMD_IS_PROFILE (self)' failed
1: 
1: (process:2250): GLib-GObject-CRITICAL **: 11:03:17.188: g_object_unref: assertion 'G_IS_OBJECT (object)' failed
1: 
1: (process:2250): GLib-GObject-CRITICAL **: 11:03:17.188: g_object_unref: assertion 'G_IS_OBJECT (object)' failed
1: .
1: (process:2250): GLib-GObject-CRITICAL **: 11:03:17.192: g_object_unref: assertion 'G_IS_OBJECT (object)' failed
1: 
1: (process:2250): libmodulemd-CRITICAL **: 11:03:17.195: modulemd_profile_get_rpms_as_strv: assertion 'MODULEMD_IS_PROFILE (self)' failed
1/4 Test #1: test_cpp .........................***Exception: SegFault  0.43 sec
```

I haven't looked at the test code, but I guess it could be caused by the lack of a NULL check before the unref in the destructor.

kontura commented:
@sbstnk sorry about that, I forgot to run the tests before creating the PR. The lack of NULL checks was indeed part of the problem, in addition to that there was also a problem with copy constructors that didn't tracked the references correctly.

sbstnk commented:
I've now built the package again with the new version of the PR and can confirm that the leak is gone. (Unfortunately the packagekitd memory usage is still very high, but this seems to be due to libsolv string interning)

lukash commented on libdnf/module/modulemd/ModuleDependencies.cpp:
```diff
@@ -27,7 +27,24 @@ ModuleDependencies::ModuleDependencies() : dependencies(nullptr) {}
+ModuleDependencies::ModuleDependencies(const ModuleDependencies& d)
+        : dependencies(d.dependencies)
+{
+    if (dependencies){
```
Missing space before `{`, same a few lines below.

lukash commented on libdnf/module/modulemd/ModuleDependencies.cpp:
```diff
@@ -27,7 +27,24 @@ ModuleDependencies::ModuleDependencies() : dependencies(nullptr) {}
+    g_object_ref(dependencies);
+}
+
+ModuleDependencies::ModuleDependencies(const ModuleDependencies& d)
```
I think we put spaces on both sides of `&` (to make it less confusing :wink: ), applies throughout the PR.

lukash commented on libdnf/module/modulemd/ModuleMetadata.cpp:
```diff
@@ -30,6 +30,14 @@ namespace libdnf {
 
+ModuleMetadata::~ModuleMetadata()
+{
+    if (resultingModuleIndex)
```
Please use scope brackets even with one-line `if`s. Same a couple lines below.

lukash commented on libdnf/module/modulemd/ModuleProfile.cpp:
```diff
@@ -25,6 +25,22 @@ namespace libdnf {
+    g_object_ref(profile);
+}
+
+ModuleProfile::ModuleProfile(ModuleProfile const& p)
```
`const ModuleProfile & p` please, first time I actually see this order :slightly_smiling_face: Same for the header file declaration.

lukash commented on libdnf/module/modulemd/ModuleProfile.cpp:
```diff
@@ -25,6 +25,22 @@ namespace libdnf {
+ModuleProfile::ModuleProfile(ModuleProfile const& p)
+        : profile(p.profile)
+{
+    if (profile){
```
A missing space here again :sunglasses:, below as well.

lukash reviewed (commented):
The rest of the review are just style nitpicks, but I'll put the important stuff here :slightly_smiling_face: 

The memory management between our wrapper classes and libmodulemd seems to be a bit of a mess.

Calling `g_object_ref()` in e.g. `ModuleDependencies(ModulemdDependencies * dependencies)` seems unintuitive and potentially dangerous to me. It works out well here (unless I missed something), because the pointer being passed is "owned" (even though it's ref-counting, hence the quotes) by something already (in particular its taken from an array inside another object). But if someone e.g. called `modulemd_dependencies_new()` and passed the result as the constructor argument, the ref count would already be 1 and `g_object_ref()` shouldn't be called.

Maybe it's actually the correct/better approach here, I may be missing the concept. But it seems unintuitive, since usually when creating objects and passing around pointers, you pass them along with ownership, (i.e. what `modulemd_dependencies_new()` does), so I find this unexpected.

Anyway, don't `ModulePackage` and `ModuleMetadata` (and perhaps others?) need a copy constructor too?

Also, when defining a copy constructor, you should define a copy assignment operator as well (see the rule of three). It is maybe not needed in the code atm., but the interface should be complete... It will create the same kind of issues if someone assumes it's good and assigns an object to another.

I can't help but think the code could use some redesign to simplify the model and make it less prone to memory errors, but I didn't study it enough and it's probably something we'll do in dnf-5 anyway.

kontura commented:
I have added the missing copy assignment operators and copy constructors. Turns out the copy assignment operators really are needed for ModuleProfiles and ModuleDependencies. :slightly_smiling_face: 

For ModulePacakge and ModuleMetadata neither of them seem to be needed and the code works fine if I `= delete` them, but I have added them in a new commit nevertheless in order to have a complete api.

I have also run the ci-dnf-stack with builds with enabled sanitizers and everything worked normally.

lukash commented on libdnf/module/ModulePackage.cpp:
```diff
@@ -85,14 +85,22 @@ static std::pair<std::string, std::string> parsePlatform(const std::string & pla
-ModulePackage::~ModulePackage() = default;
+ModulePackage::~ModulePackage()
+{
+    if (mdStream) {
```
Sorry about not mentioning this the first time, but it's better to check pointers explicitly (for clarity of coding style, it works the same, of course): `if (mdStream != nullptr)`

There's a lot of those throughout the PR...

lukash commented on libdnf/module/modulemd/ModuleDependencies.hpp:
```diff
@@ -36,6 +36,10 @@ class ModuleDependencies
     explicit ModuleDependencies();
     explicit ModuleDependencies(ModulemdDependencies *dependencies);
+    ModuleDependencies & operator=(const ModuleDependencies & d);
+    ModuleDependencies(const ModuleDependencies & d);
```
Really nitpicky, but I prefer to use the same order of these consistently:
```
// constructors
// assignment operator(s)
// destructor
```

lukash commented on libdnf/module/ModulePackage.cpp:
```diff
@@ -113,6 +113,32 @@ ModulePackage::ModulePackage(DnfSack * moduleSack, LibsolvRepo * repo,
+
+ModulePackage & ModulePackage::operator=(const ModulePackage & mpkg)
+{
+    if (mdStream) {
```
This one could use the `if (this != &mpkg)` check too.

lukash reviewed (commented):
> I have also run the ci-dnf-stack with builds with enabled sanitizers and everything worked normally.

Out of curiosity, did you try to run the test suite with sanitizers and some of the issues this PR is solving unfixed, to see if the sanitizers would catch it? I'd be really interested to know whether they would...

kontura committed:
Add unused copy constructors and assignment operators for complete API

None of these are actually used anywhere right now, but add them
in order to have more complete and working API for Modular wrappers.

kontura commented:
I have pushed an updated version. Thanks for the review!

I did some testing with the sanitizers on (not the ci-dnf-stack because there is a lot of output to stdout) but wasn't able to see any difference when running through DNF, the same applies to valgrind. When running through packagekit its possible to see the difference in valgrind outputs (the memory leak described in the original issue is gone with this patch).

I have also tried getting the libdnf sanitizers output when called from packagekit but wasn't able to get it, it is probably eaten by the packagekit daemon somewhere.

lukash commented:
@rh-atomic-bot r+
